GH-1704 Broader Batch/RecordInterceptor

kdowbecki opened the pull request.

pivotal-cla commented:
@kdowbecki Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=1912)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=1912) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@kdowbecki Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=1912)!

kdowbecki commented on spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java:
```diff
@@ -1233,6 +1233,9 @@ public void run() {
 					handleConsumerException(e);
 				}
+				finally {
+					finishInvoke();
```
I wanted it to run after `handleConsumerException()` so the trace context will be available for an exception outside of the listener processing e.g. when offsets fail to commit.

garyrussell replied:
Right, but as I said in my comment on the issue; it needs to be called conditionally in the finally block - only if an exception was thrown; otherwise it will be called twice.

kdowbecki replied:
You said:

> We need to handle the `poll()` throwing an exception, with a conditional call to `finishInvoke()` in a `finally` block in the `run()` method. e.g. a boolean `success` set to false just after the `while()` and true after the `pollAndInvoke()` and use it in the `finally` block to call the clean up if an exception occurred.

My understanding of the tracing use-case that started this PR is that `finishInvoke()` must be run if `beforePoll()` was run. Whether or not records were returned from `poll()`, or processed successfully by listener, or exception was thrown somewhere, is not important. The only goal is to run `finishInvoke()` after `beforePoll()`.

Can you please explain what do you mean by:

>  clean up if an exception occurred.

because it feels like the `finishInvoke()` cleanup is not dependent on an exception being thrown during processing. Unless you are suggesting that `finishInvoke()` should only run if `beforePoll()` run without throwing an exception? My apologies, I'm clearly missing something.

garyrussell replied:
Well, I was thinking that we should call `finishInvoke()` after each record in a record listener but I guess you are right; these methods should be just scoped around the `poll()` and completion of processing all records from the poll.

In that case, we should rename `finishInvoke` to something like `recordsProcessed()` to decouple it from invoking the listener.

Most likely record interceptors won't use them, but we should provide them for consistency.

kdowbecki replied:
I renamed `finishInvoke()` to `afterRecordsProcessed()` and improved the Javadoc. 

I still have to figure out how to test it because `beforePoll()` and `afterRecordsProcessed()` is called in a loop outside of the test. I plan to look into it over the weekend.

kdowbecki commented on spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java:
```diff
@@ -1441,6 +1476,37 @@ private void fixTxOffsetsIfNeeded() {
+					this.commonBatchInterceptor.beforePoll(this.consumer);
+				}
+				catch (Exception e) {
+					this.logger.error(e, "BatchInterceptor threw an exception");
```
Do we need a more detailed log message to tell us that it happened for `beforePoll()`. Right now the same message is duplicated in few places for `beforePoll()`, `intercept()` and `finishInvoke()`. It feels like exceptions when invoking interceptor are a corner case, not sure if anyone really pays attention to this log.

garyrussell replied:
The stack trace is enough to tell us where it happened, but it wouldn't hurt to mention the method name; interceptors should not throw exceptions, but we need to protect against it.

garyrussell commented:
I am not working today - I will try to do a full review tomorrow.

kdowbecki commented on spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java:
```diff
@@ -1270,6 +1273,7 @@ protected void pollAndInvoke() {
 				return;
 			}
 			this.polling.set(true);
+			beforePoll();
```
`pollAndInvoke()` starts with offset related operation e.g. it can commit offsets. Are these offsets of the previous record/batch processing? Perhaps it would be better to call `finishInvoke()` just before `idleBetweenPollIfNecessary()` few lines above if single batch of records is fully processed in two `pollAndInvoke()` invocations?

garyrussell replied:
I concur that `beforePoll()` should be imediately before the poll. It's not clear why you have 2 calls to `beforePoll()` (one here and on in each branch of `doPoll()`.

However, we now have a new method `pollConsumer()` (common to both batch and record listeners - the `beforePoll()` should be done there.

I don't think that calling `afterRecordsProcessed()` before `processCommits()` will cause any problems; also, moving it from the `finally` block might cause us to miss the last call (when `isRunning()` goes false, when the container is stopped.

kdowbecki replied:
Two calls to `beforePoll()` method was bug. I removed redundant call when writing `InOrder` tests.

garyrussell replied:
Thanks - I will take another look tomorrow; thanks again.

kdowbecki commented on spring-kafka/src/test/java/org/springframework/kafka/annotation/EnableKafkaIntegrationTests.java:
```diff
@@ -178,9 +180,10 @@
 		"annotated18", "annotated19", "annotated20", "annotated21", "annotated21reply", "annotated22",
 		"annotated22reply", "annotated23", "annotated23reply", "annotated24", "annotated24reply",
 		"annotated25", "annotated25reply1", "annotated25reply2", "annotated26", "annotated27", "annotated28",
-		"annotated29", "annotated30", "annotated30reply", "annotated31", "annotated32", "annotated33",
```
`annotated29` was not used as far as I can tell.

kdowbecki commented on spring-kafka/src/test/java/org/springframework/kafka/annotation/EnableKafkaIntegrationTests.java:
```diff
@@ -600,7 +603,16 @@ public void testBatchRecords() throws Exception {
+	@Test
+	public void testBatchInterceptor() throws Exception {
+		template.send("annotated43", null, "foo");
+		assertThat(this.config.batchAfterRecordsProcessedLatch.await(30, TimeUnit.SECONDS)).isTrue();
```
Because I can't wait on a latch inside `@KafkaListener`, it's too early and the test might run next assertion before the interceptor methods are called. Current approach doesn't work because `batchAfterRecordsProcessedLatch` is called too early, in the previous `pollAndInvoke()` iteration before the actual test sends the message. I need to improve it.

kdowbecki commented:
@garyrussell Apologies for delay, I fixed the tests. I haven't found a way to test it perfectly because `poll()` is done in a loop, I can't clearly distinguish one `poll()` invocation from another without writing a lot of test code. I didn't went this route and written a simple test. Can you please review?

garyrussell reviewed (changes requested):
For the test scenario, we would normally use mocks an an `InOrder` to ensure the consumer.poll and interceptor methods are called in the right sequence.

There are several examples in other test cases; if you are not comfortable with doing that, I can add a test before the final merge.

Thanks again for the contribution!

kdowbecki commented:
I added new `InOrder` interceptor tests in `KafkaMessageListenerContainerTests` class. I wonder do we still need the integration tests that I added before in this PR in `EnableKafkaIntegrationTests` class. I'm leaning towards "no" since these old tests verify the same thing, I think I should revert `EnableKafkaIntegrationTests`. What's your opinion @garyrussell ?

garyrussell commented:
@kdowbecki Nice work on the mock tests; thanks; yes, I agree, we can revert the integration tests.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/listener/BatchInterceptor.java:
```diff
@@ -61,4 +61,29 @@ default void success(ConsumerRecords<K, V> records, Consumer<K, V> consumer) {
+	 * @param consumer the consumer.
+	 * @since 2.8
+	 */
+	default void afterRecordsProcessed(Consumer<K, V> consumer) {
```
Don't we need to have a `ConsumerRecords<K, V> records` context over here as well?
You maybe would like to know for what batch is this stateless interceptor.
See an `ExecutorChannelInterceptor` contract in Spring Messaging for example.

garyrussell replied:
That's a good suggestion.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/listener/RecordInterceptor.java:
```diff
@@ -81,4 +81,29 @@ default void success(ConsumerRecord<K, V> record, Consumer<K, V> consumer) {
+	 * @param consumer the consumer.
+	 * @since 2.8
+	 */
+	default void afterRecordsProcessed(Consumer<K, V> consumer) {
```
This one is for a single record interceptor, so the contract must be called `afterRecordProcessed` and have a `ConsumerRecord<K, V> record` as a first argument.
I might miss something and probably this per-record interceptor is not designed for an "after processed" hook...
Otherwise it sounds awkward to implement a `RecordInterceptor` and figure out somehow a contract for the whole batch as it sounds right now.

garyrussell replied:
No; this would be redundant - we already call success or failure for each record.

This hook is to clean up state when all records have been processed - the `beforePoll()` is intended so that thread-bound resources can be registered and made available in components called by the `KafkaConsumer` within the `poll()` (e.g. deserializers, partition assignors, etc, etc).

This hook is to unbind those resources.

artembilan replied:
Right; so should not be there then some other contract not the one tied to the interceptors?
It feel right now for me that this is not an interception  logic: we just don't intercept anything before and after.
Feels more like a `BeforeAfterPollProcessor`.
No? What was a premise to put such a logic into the interceptor contract?

garyrussell replied:
It was [my suggestion on the issue](https://github.com/spring-projects/spring-kafka/issues/1704#issuecomment-896997190).

However, I see that we could.should extract a super-interface with these two methods and have the interceptors inherit from it.

The challenge is naming it 😄 

Suggestions?

garyrussell replied:
It was [my suggestion on the issue](https://github.com/spring-projects/spring-kafka/issues/1704#issuecomment-896997190).

However, I see that we could/should extract a super-interface with these two methods and have the interceptors inherit from it.

The challenge is naming it 😄 

`BeforeAfterPollProcessor` sounds ok to me.

artembilan replied:
That will work for me, too.
So, the logic in the Listener Container would be much simpler.

kdowbecki replied:
Works for me, I'll add `BeforeAfterPollProcessor`

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java:
```diff
@@ -1455,6 +1490,37 @@ private void fixTxOffsetsIfNeeded() {
+			}
+			else {
+				recordInterceptBeforePoll();
+			}
```
Sounds more like there has to be a common Before-After contract, but not the same copy in record and batch interceptors.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java:
```diff
@@ -1236,6 +1236,9 @@ public void run() {
 					handleConsumerException(e);
 				}
+				finally {
+					afterRecordsProcessed();
```
I see now why no records in the "after" contract. 
Probably that `pollAndInvoke()` may return that `ConsumerRecords`?
Yes, it might be a `null` since we may fail doing all of that hard logic, but still might be better to propagate such a context down to this "after" contract invocation.

garyrussell replied:
I guess the same comment about redundancy applies here too - the batch interceptor gets the `ConsumerRecord` ini success or failure and the record interceptor gets each record individually.

kdowbecki replied:
Perhaps the issue is naming of `afterRecordsProcessed()` method. Perhaps we should remove `records` from it and call it `afterPollProcessed()` or similar?

artembilan replied:
If we go with a new `BeforeAfterPollProcessor` contract, then just `afterPoll()` would be enough. 
Something in parallel with that `beforePoll()`.

garyrussell replied:
👍 

I am planning on building the next milestone (2.8.0-M3) later today; if would be great if we can get this into that ready for the next Boot milestone next week.

It's not really due until Monday, but I will be on PTO next week.

garyrussell replied:
I don't think `afterPoll()` sounds right; it implies it's done right after. `afterPollProcessed()` sounds good.

garyrussell replied:
or `afterPollResultsProcessed()`.

artembilan replied:
OK. Then I'm good with that `afterPollProcessed()`.

kdowbecki replied:
`AfterRollbackProcessor` and few other classes have `clearThreadState()` method already. Maybe instead of `afterPoll()` it should be `clearThreadState()` hear as well?

garyrussell replied:
I could live with `clearThreadState()`.

I am not sure `ConsumerPollProcessor` conveys the right meaning.

(I agree that naming is hard 😄 ).

Still thinking - let's stick with `BeforeAfterPollProcessor.clearThreadState()` for now.

It's only a milestone, so we are at liberty to change it if we can come up with something better.

...time passes...

How about something more general...

```java
public ThreadScopeProcessor {

    void setupThreadState();

   void clearThreadState();

}
```

??

kdowbecki replied:
Unfortunately it seems that we won't be able to do `AfterRollbackProcessor extends ThreadScopeProcessor` because `AfterRollbackProcessor` lacks the setup side. I like the generic `ThreadScopeProcessor` as an idea but it would have to be applied in more place to be valid. If it's just this one place I like other names that mention "poll" better.

I finished refactor the code and used `BeforeAfterPollProcessor`. It's getting late here in London so I will leave the renaming of the classes to you. I'm happy with either option.

garyrussell replied:
I think we could make the `setUpThreadState` a default method; the DARP and stateful error handlers set up thread state as a side-effect of their normal processing.

garyrussell replied:
Thanks for the work - I can take it from here.

kdowbecki replied:
Few minor fixed to compiler warnings, but now I'm done. Have fun, and thanks for helping me with the PR.

artembilan commented on spring-kafka/src/test/java/org/springframework/kafka/listener/KafkaMessageListenerContainerTests.java:
```diff
@@ -3497,6 +3497,280 @@ void stopImmediately() throws InterruptedException {
+		records.put(new TopicPartition("foo", 0), List.of(firstRecord, secondRecord));
+		ConsumerRecords<Integer, String> consumerRecords = new ConsumerRecords<>(records);
+		given(consumer.poll(any(Duration.class))).willAnswer(i -> {
+			Thread.sleep(50);
```
Do we really need this sleep?
Not clear what further logic expects some delays with records delivery...

garyrussell replied:
Yes; otherwise the consumer thread will call it in a tight loop.

artembilan replied:
OK. won't another `willAnswer(i -> null)` after this one help?
According Mockito docs only the last stub is used for the rest of call.
We probably need to have a idle on the container in addition...
But yeah, we can live it as is: just wanted to share with you a logic possible via `.willAnswer(...).willAnswer(...)` chain.

garyrussell replied:
No; returning null will not prevent the tight loop; we could add a latch so the second poll blocks until the test completes but I don't feel it's that important - there are many tests with this logic; we can put in on the slate for improvement later.

artembilan reviewed (changes requested):
I might be late for the party and may miss something you have discussed before, but I thought it won't hurt too much if I raise my concerns even if they won't be accepted in the end.

Thank you!

garyrussell commented:
Thanks again for this; I will submit a new PR with the polishing we discussed.
